### ISSUE TITLE
False Negative: `abstract`, `native`, and signature-edge cases are not treated as real missing-pair violations.

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6535]
# False Negative: `abstract`, `native`, and signature-edge cases are not treated as real missing-pair violations.

Version
7.22.0

## Checker
- Checker id: `category/java/errorprone.xml/OverrideBothEqualsAndHashcode`
- Checker description: Detects when a class (including anonymous classes and records) overrides either `equals` or `hashCode` but not both, unless the class implements `Comparable`.

## Description of the false negative
`diff/output` marks 2 false negatives for `category/java/errorprone.xml/OverrideBothEqualsAndHashcode`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The recurring themes are abstract declarations, native declarations.

### `PosCase5.java`
Source: `output/checkstyle/EqualsHashCodeCheck/src/main/java/scensct/core/pos/PosCase5.java`

This class only declares one side of the `equals`/`hashCode` pair abstractly instead of providing a concrete implementation. The missed report suggests the implementation is treating abstract declarations as if they fully satisfied the rule's counterpart requirement. That is also how the benchmark classifies the sample: A class defines a concrete equals(Object) method and any zero-parameter hashCode method defined is abstract should be flagged as violation.

```java
// A class defines a concrete equals(Object) method and any zero-parameter hashCode method defined is abstract should be flagged as violation.
package scensct.core.pos;

public abstract class PosCase5 {
    private int value;

    // Concrete equals method defined
    public boolean equals(Object obj) { // [REPORTED LINE]
        if (obj instanceof PosCase5) {
            return this.value == ((PosCase5) obj).value;
        }
        return false;
    }

    // Abstract hashCode method (no body, not native)
    public abstract int hashCode();
}
```

### `PosCase6.java`
Source: `output/checkstyle/EqualsHashCodeCheck/src/main/java/scensct/core/pos/PosCase6.java`

This class only declares one side of the `equals`/`hashCode` pair abstractly instead of providing a concrete implementation. The missed report suggests the implementation is treating abstract declarations as if they fully satisfied the rule's counterpart requirement. That is also how the benchmark classifies the sample: A class defines a native equals(Object) method and any zero-parameter hashCode method defined is abstract should be flagged as violation.

```java
// A class defines a native equals(Object) method and any zero-parameter hashCode method defined is abstract should be flagged as violation.
package scensct.core.pos;

public abstract class PosCase6 {
    private int value;

    // Native equals method defined
    public native boolean equals(Object obj); // [REPORTED LINE]

    // Abstract hashCode method (no body, not native)
    public abstract int hashCode();
}
```

## Cause analysis
The rule is documented as follows: Detects when a class (including anonymous classes and records) overrides either `equals` or `hashCode` but not both, unless the class implements `Comparable`.

In this set the problem shows up around abstract declarations, native declarations. The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `category/java/errorprone.xml/OverrideBothEqualsAndHashcode`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.

**Comment by oowekyala** [comment]
This doesn't look like a false negative to me. The hashcode or equals implementation has not been forgotten as it is implemented in native code or by subclasses.
